File: swift/leet_code/463.swift
/**
 An `Solution` class that calculates the perimeter of an island in a grid.

 The grid is represented as a 2D array, where 0 represents water and 1 represents land. The grid cells are connected horizontally and vertically but not diagonally. The perimeter of the island is the total number of land-to-water or land-to-grid-edge transitions.

 - Note: The grid is assumed to be rectangular, with the same number of columns in each row.
 */
import Foundation

class Solution {
    /**
     Calculates the perimeter of the island in the given grid.

     This method traverses the grid from top to bottom and left to right. For each land cell, it adds 4 to the perimeter (since a standalone cell has 4 sides). Then, if the cell has a neighbor to the right or down that is also a land cell, it subtracts 2 from the perimeter (since each shared side between two cells reduces the total perimeter by 2).

     - Parameter grid: A 2D array representing the grid. Each element is either 0 (water) or 1 (land).
     - Returns: The total perimeter of the island in the grid.
     - Throws: This method does not throw any error.
     - Precondition: The grid is non-empty and rectangular, with the same number of columns in each row.
     - Postcondition: The returned value is non-negative and represents the total perimeter of the island in the grid.

     - Example:
     ```
     let solution = Solution()
     let grid = [[0, 1, 0, 0], [1, 1, 1, 0], [0, 1, 0, 0], [1, 1, 0, 0]]
     let perimeter = solution.islandPerimeter(grid)
     print(perimeter)  // Prints "16"
     ```
     */
    func islandPerimeter(_ grid: [[Int]]) -> Int {
        let rows = grid.count
        let columns = grid[0].count

        var perimeter = 0

        for i in 0 ..< rows {
            for j in 0 ..< columns {
                if grid[i][j] == 1 {
                    perimeter += 4

                    // Check the cell below
                    if i < rows - 1, grid[i + 1][j] == 1 {
                        perimeter -= 2
                    }

                    // Check the cell to the right
                    if j < columns - 1, grid[i][j + 1] == 1 {
                        perimeter -= 2
                    }
                }
            }
        }

        return perimeter
    }
}

File: scala/leet_code/463.scala
import scala.collection.mutable.ArrayBuffer

/** This is the Solution object. It contains a method to calculate the perimeter of an island in a given 2D array. Each
  * cell in the 2D array represents a piece of land, where 1 represents land and 0 represents water. The method returns
  * an integer representing the total perimeter of the island.
  *
  * @example
  *   val grid = Array(Array(0,1,0,0), Array(1,1,1,0), Array(0,1,0,0), Array(1,1,0,0)) val result =
  *   Solution.islandPerimeter(grid) // result: 16
  *
  * @note
  *   The method does not modify the input array.
  */
object Solution {

  /** Calculates the perimeter of an island in the given 2D array.
    *
    * @param grid
    *   A 2D array representing the grid. Each cell represents a piece of land, where 1 represents land and 0 represents
    *   water.
    * @return
    *   An integer representing the total perimeter of the island.
    * @throws IllegalArgumentException
    *   if the input array is empty or not rectangular.
    */
  def islandPerimeter(grid: Array[Array[Int]]): Int = {
    // Check if the input array is empty or not rectangular
    if (grid.isEmpty || grid.exists(_.length != grid(0).length)) {
      throw new IllegalArgumentException(
        "Input array must be non-empty and rectangular"
      )
    }

    // Helper function to count the perimeter of a land cell
    def countPerimeter(i: Int, j: Int): Int = {
      var perimeter: Int = 4
      if (i - 1 >= 0 && grid(i - 1)(j) == 1) perimeter -= 1
      if (i + 1 < grid.length && grid(i + 1)(j) == 1) perimeter -= 1
      if (j - 1 >= 0 && grid(i)(j - 1) == 1) perimeter -= 1
      if (j + 1 < grid(i).length && grid(i)(j + 1) == 1) perimeter -= 1
      perimeter
    }

    var totalPerimeter: Int = 0

    // Iterate over each cell in the array
    for (i <- grid.indices) {
      for (j <- grid(i).indices) {
        // If the cell represents land
        if (grid(i)(j) == 1) {
          // Add the perimeter of the land cell to the total perimeter
          totalPerimeter += countPerimeter(i, j)
        }
      }
    }

    // Return the total perimeter
    totalPerimeter
  }
}
